File: js/repl/lib/src/Promise.js
import { mapError, mapOk, Result } from "../fable-library.2.4.2/Option.js";
import { declare } from "../fable-library.2.4.2/Types.js";
import { type } from "../fable-library.2.4.2/Reflection.js";
import { iterate } from "../fable-library.2.4.2/Seq.js";
import { equals } from "../fable-library.2.4.2/Util.js";
export function reject(reason) {
  return Promise.reject(reason);
}
export function result(a) {
  return a.then(function ($arg$$1) {
    let arg0;
    arg0 = new Result(0, "Ok", $arg$$1);
    return arg0;
  }, function ($arg$$2) {
    let arg0$$2;
    arg0$$2 = new Result(1, "Error", $arg$$2);
    return arg0$$2;
  });
}
export function mapResult(fn, a$$1) {
  return a$$1.then(function a$$2(result$$1) {
    return mapOk(fn, result$$1);
  });
}
export function bindResult(fn$$1, a$$3) {
  return a$$3.then(function a$$6(a$$4) {
    if (a$$4.tag === 1) {
      return Promise.resolve(new Result(1, "Error", a$$4.fields[0]));
    } else {
      return result(fn$$1(a$$4.fields[0]));
    }
  });
}
export function mapResultError(fn$$2, a$$7) {
  return a$$7.then(function a$$8(result$$2) {
    return mapError(fn$$2, result$$2);
  });
}
export function tap(fn$$3, a$$9) {
  return a$$9.then(function a$$10(x) {
    fn$$3(x);
    return x;
  });
}
export const PromiseBuilder = declare(function Promise_PromiseBuilder() {});
export function PromiseBuilder$reflection() {
  return type("Promise.PromiseBuilder");
}
export function PromiseBuilder$$$$002Ector() {
  return this instanceof PromiseBuilder ? PromiseBuilder.call(this) : new PromiseBuilder();
}
export function PromiseBuilder$$For$$1565554B(x$$1, seq, body) {
  let p = Promise.resolve(null);
  iterate(function (a$$11) {
    const x$$2 = p.then(function () {
      return body(a$$11);
    });
    p = x$$2;
  }, seq);
  return p;
}
export function PromiseBuilder$$While$$2044D34(x$$3, guard, p$$1) {
  if (guard()) {
    return p$$1.then(function () {
      return PromiseBuilder$$While$$2044D34(x$$3, guard, p$$1);
    });
  } else {
    return Promise.resolve(null);
  }
}
export function PromiseBuilder$$TryFinally$$7D49A2FD(x$$4, p$$2, compensation) {
  return p$$2.then(function (x$$5) {
    compensation();
    return x$$5;
  }, function (er) {
    compensation();
    throw er;
  });
}
export function PromiseBuilder$$Delay$$62FBFDE1(x$$7, generator) {
  const x$$13 = {
    then(f1, f2) {
      try {
        return generator().then(f1, f2);
      } catch (er$$1) {
        if (equals(f2, null)) {
          const x$$8 = Promise.reject(er$$1);
          return x$$8;
        } else {
          try {
            const x$$9 = Promise.resolve(f2(er$$1));
            return x$$9;
          } catch (er$$2) {
            const x$$10 = Promise.reject(er$$2);
            return x$$10;
          }
        }
      }
    },

    catch(f) {
      try {
        return generator().catch(f);
      } catch (er$$3) {
        try {
          const x$$11 = Promise.resolve(f(er$$3));
          return x$$11;
        } catch (er$$4) {
          const x$$12 = Promise.reject(er$$4);
          return x$$12;
        }
      }
    }

  };
  return x$$13;
}
export function PromiseBuilder$$Run$$212F1D4B(x$$14, p$$3) {
  return new Promise(function (success, fail) {
    try {
      let p$$4;
      const x$$15 = Promise.resolve(p$$3);
      p$$4 = x$$15;
      p$$4.then(success, fail);
    } catch (er$$5) {
      fail(er$$5);
    }
  });
}
export function PromiseBuilder$$Using$$Z7FDC6BE3(x$$16, resource, binder) {
  return PromiseBuilder$$TryFinally$$7D49A2FD(x$$16, binder(resource), function () {
    let copyOfStruct = resource;
    copyOfStruct.Dispose();
  });
}

File: src/Fable.Repl.Lib/Promise/Promise.fs

[<RequireQualifiedAccess>]
module Promise

#nowarn "1182" // Unused values

open System
open Fable.Core
open Fable.Core.JsInterop

[<Emit("new Promise($0)")>]
/// The promise function receives two other function parameters: success and fail
let create (f: ('T -> unit) -> (exn -> unit) -> unit): JS.Promise<'T> = jsNative

[<Emit("new Promise(resolve => setTimeout(resolve, $0))")>]
let sleep (ms: int): JS.Promise<unit> = jsNative

[<Emit("Promise.resolve($0)")>]
let lift<'T> (a: 'T): JS.Promise<'T> = jsNative

[<Emit("Promise.reject($0)")>]
/// Creates promise (in rejected state) with supplied reason.
let reject<'T> (reason: exn) : JS.Promise<'T> = jsNative

[<Emit("$1.then($0)")>]
let bind (a: 'T1 -> JS.Promise<'T2>) (pr: JS.Promise<'T1>): JS.Promise<'T2> = jsNative

[<Emit("$1.then($0)")>]
let map (a: 'T1 -> 'T2) (pr: JS.Promise<'T>): JS.Promise<'T2> = jsNative

[<Emit("void ($1.then($0))")>]
let iter (a: 'T -> unit) (pr: JS.Promise<'T>): unit = jsNative

let tap (fn: 'T -> unit) (a: JS.Promise<'T>): JS.Promise<'T> =
    a |> map (fun x -> fn x; x)

[<Emit("$1.catch($0)")>]
/// This version of `catch` expects a function returning just 'T, as opposed to `Promise<'T>`. If you need to return `Promise<'T>`, use `catchBind`.
let catch (fail: exn -> 'T) (pr: JS.Promise<'T>): JS.Promise<'T> = jsNative

[<Emit("$1.catch($0)")>]
/// This version of `catch` expects a function returning Promise<'T> as opposed to just 'T. If you need to return just 'T, use `catch`.
let catchBind (fail: exn -> JS.Promise<'T>) (pr: JS.Promise<'T>): JS.Promise<'T> = jsNative

[<Emit("void ($1.catch($0))")>]
/// Used to catch errors at the end of a promise chain.
let catchEnd (fail: exn -> unit) (pr: JS.Promise<'T>): unit = jsNative

[<Emit("$2.then($0).catch($1)")>]
/// A combination of `map` and `catch`, this function applies the `success` continuation when the input promise resolves successfully, or `fail` continuation when the input promise fails.
let either (success: 'T1 -> 'T2) (fail: exn -> 'T2) (pr: JS.Promise<'T1>): JS.Promise<'T2> = jsNative

[<Emit("$2.then($0).catch($1)")>]
/// A combination of `bind` and `catchBind`, this function applies the `success` continuation when the input promise resolves successfully, or `fail` continuation when the input promise fails.
let eitherBind (success: 'T1 -> JS.Promise<'T2>) (fail: exn -> JS.Promise<'T2>) (pr: JS.Promise<'T1>): JS.Promise<'T2> = jsNative

[<Emit("void ($2.then($0).catch($1))")>]
let eitherEnd (success: 'T -> unit) (fail: exn -> unit) (pr: JS.Promise<'T>): unit = jsNative

[<Emit("void $0")>]
let start (pr: JS.Promise<'T>): unit = jsNative

[<Emit("$1.catch($0)")>]
let tryStart (fail: exn -> unit) (pr: JS.Promise<'T>): unit = jsNative

[<Emit("Promise.all($0)")>]
let Parallel (pr: seq<JS.Promise<'T>>): JS.Promise<'T[]> = jsNative

[<Emit("Promise.all($0)")>]
let all (pr: seq<JS.Promise<'T>>): JS.Promise<'T[]> = jsNative

let result (a: JS.Promise<'T>): JS.Promise<Result<'T, exn>> =
    either Ok Error a

let mapResult (fn: 'T1 -> 'T2) (a: JS.Promise<Result<'T1, 'E>>): JS.Promise<Result<'T2, 'E>> =
    a |> map (Result.map fn)

let bindResult (fn: 'T1 -> JS.Promise<'T2>) (a: JS.Promise<Result<'T1, 'E>>): JS.Promise<Result<'T2, 'E>> =
    a |> bind (fun a ->
        match a with
        | Ok a ->
            fn a |> map Ok
        | Error e ->
            lift (Error e))

let mapResultError (fn: 'E1 -> 'E2) (a: JS.Promise<Result<'T, 'E1>>): JS.Promise<Result<'T, 'E2>> =
    a |> map (Result.mapError fn)

type PromiseBuilder() =
    [<Emit("$1.then($2)")>]
    member _.Bind(p: JS.Promise<'T1>, f: 'T1 -> JS.Promise<'T2>): JS.Promise<'T2> = jsNative

    [<Emit("$1.then(() => $2)")>]
    member _.Combine(p1: JS.Promise<unit>, p2: JS.Promise<'T>): JS.Promise<'T> = jsNative

    member _.For(seq: seq<'T>, body: 'T -> JS.Promise<unit>): JS.Promise<unit> =
        // (lift (), seq)
        // ||> Seq.fold (fun p a ->
        //     bind (fun () -> body a) p)
        let mutable p = lift ()
        for a in seq do
            p <- p |> bind (fun () -> body a)
        p

    [<Emit("$1.then($2)")>]
    member _.For(p: JS.Promise<'T1>, f: 'T1 -> JS.Promise<'T2>): JS.Promise<'T2> = jsNative

    member this.While(guard: unit -> bool, p: JS.Promise<unit>): JS.Promise<unit> =
        if guard()
        then bind (fun () -> this.While(guard, p)) p
        else lift()

    [<Emit("Promise.resolve($1)")>]
    member _.Return(a: 'T): JS.Promise<'T> = jsNative

    [<Emit("$1")>]
    member _.ReturnFrom(p: JS.Promise<'T>): JS.Promise<'T> = jsNative

    [<Emit("Promise.resolve()")>]
    member _.Zero(): JS.Promise<unit> = jsNative

    member _.TryFinally(p: JS.Promise<'T>, compensation: unit -> unit): JS.Promise<'T> =
        either (fun (x: 'T) -> compensation(); x) (fun er -> compensation(); raise er) p

    [<Emit("$1.catch($2)")>]
    member _.TryWith(p: JS.Promise<'T>, catchHandler: exn -> JS.Promise<'T>): JS.Promise<'T> = jsNative

    // Delay must generate a cold promise-like object that re-runs every time it's called,
    // so we cannot use the JS Promise constructor which is stateful
    member _.Delay(generator: unit -> JS.Promise<'T>): JS.Promise<'T> =
        !!createObj[
            "then" ==> fun onSuccess onError ->
                try generator().``then``(onSuccess, onError)
                with er ->
                    if isNull(box onError) then reject er
                    else
                        try onError er |> lift
                        with er -> reject er
            "catch" ==> fun onError ->
                try generator().catch(onError)
                with er ->
                    try onError er |> lift
                    with er -> reject er
        ]

    // Make sure we call `then` because this may be used with "cold" fake promises generated by Delay
    member _.Run(p:JS.Promise<'T>): JS.Promise<'T> = p.``then``(id)

    member this.Using<'T1, 'T2 when 'T1 :> IDisposable>(resource: 'T1, binder: 'T1 -> JS.Promise<'T2>): JS.Promise<'T2> =
        this.TryFinally(binder(resource), fun () -> resource.Dispose())

    [<Emit("Promise.all([$1, $2])")>]
    member _.MergeSources(a: JS.Promise<'T1>, b: JS.Promise<'T2>): JS.Promise<'T1 * 'T2> = jsNative

    [<Emit("Promise.all([$1, $2, $3])")>]
    member _.MergeSources3(a: JS.Promise<'T1>, b: JS.Promise<'T2>, c: JS.Promise<'T3>): JS.Promise<'T1 * 'T2 * 'T3> = jsNative

    [<Emit("Promise.all([$1, $2, $3, $4])")>]
    member _.MergeSources4(a: JS.Promise<'T1>, b: JS.Promise<'T2>, c: JS.Promise<'T3>, d: JS.Promise<'T4>): JS.Promise<'T1 * 'T2 * 'T3 * 'T4> = jsNative

    [<Emit("Promise.all([$1, $2, $3, $4, $5])")>]
    member _.MergeSources5(a: JS.Promise<'T1>, b: JS.Promise<'T2>, c: JS.Promise<'T3>, d: JS.Promise<'T4>, e: JS.Promise<'T5>): JS.Promise<'T1 * 'T2 * 'T3 * 'T4 * 'T5> = jsNative

    [<Emit("Promise.all([$1, $2, $3, $4, $5, $6])")>]
    member _.MergeSources6(a: JS.Promise<'T1>, b: JS.Promise<'T2>, c: JS.Promise<'T3>, d: JS.Promise<'T4>, e: JS.Promise<'T5>, f: JS.Promise<'T6>): JS.Promise<'T1 * 'T2 * 'T3 * 'T4 * 'T5 * 'T6> = jsNative

//    member _.BindReturn(y: JS.Promise<'T1>, f) = map f y

    [<Emit("Promise.all([$1,$2]).then(([a,b]) => $3(a,b))")>]
    [<CustomOperation("andFor", IsLikeZip=true)>]
    member _.Merge(a: JS.Promise<'T1>, b: JS.Promise<'T2>, [<ProjectionParameter>] resultSelector : 'T1 -> 'T2 -> 'R): JS.Promise<'R> = jsNative

    /// this is the 'base case' for Source transformations, and
    /// must be present before any user-defined overloads will
    /// actually work.
    member x.Source(p: JS.Promise<'T1>): JS.Promise<'T1> = p

    // allows for..in and for..do expressions inside CEs
    member x.Source(ps: #seq<_>): _ = ps
